File: account-management-server/src/main/java/com/ts/account_management_server/model/database/UserDevice.java
package com.ts.account_management_server.model.database;

import com.ts.account_management_server.model.enums.UserDeviceRole;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import java.util.Collection;
import java.util.List;
import java.util.Map;

@NoArgsConstructor
@AllArgsConstructor
@Data
@Document(collection = "user-devices")
public class UserDevice {
    @Id
    private String id;

    @Indexed(unique = true)
    private String label;

    private UserDeviceRole role;

    private String currentToken;

    private Map<String, String> metadata;

    @Indexed(unique = true)
    private String secretKey;

    private int allowedTokenRequestCount;

    public Collection<? extends GrantedAuthority> getAuthorities() {
        return List.of(new SimpleGrantedAuthority(role.name()));
    }
}

File: account_management_client/lib/model/entity/user_device.dart
class UserDevice {
  String id;
  String role;
  String label;
  Map<String, String> metadata;
  int allowedTokenRequestCount;

  UserDevice(
      {required this.id,
      required this.role,
      required this.label,
      required this.metadata,
      required this.allowedTokenRequestCount});

  factory UserDevice.fromJson(Map<String, dynamic> json) {
    return UserDevice(
        id: json['id'],
        role: json['role'],
        label: json['label'],
        metadata: Map<String, String>.from(json['metadata']),
        allowedTokenRequestCount: json['allowedTokenRequestCount']);
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'role': role,
      'label': label,
      'metadata': metadata,
      'allowedTokenRequestCount': allowedTokenRequestCount,
    };
  }
}
